3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,202 @@
+     * @private
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * @type {!Array<Object>}
+     * @private
+     */
+    this.receivedExtraDataQueue_ = [];
+
+    /**
+     * Holds a mapping between sender ID and extra data
+     * @type {Map}
+     * @private
+     */
+    this.senderIdToExtraData_ = map();
+
+    /**
+     * Simple requestIdleCallback polyfill
+     * @type {Function<Function>}
+     * @private
+     */
+    this.requestIdleCallback_ =
+      win.requestIdleCallback.bind(win) || (cb => setTimeout(cb, 1));
+
+    /**
+     * Handles communication between frames
+     * @type {IframeMessagingClient}
+     * @private
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(JSON.parse(window.name).sentinel);
+    this.iframeMessagingClient_.registerCallback('ampAnalytics3pMessages',
+      this.onReceivedMessagesFromCreative_.bind(this));
+
+    this.sendReadyMessageToCreative_();
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur
+   * @param {!Function} listener A function that takes an array of event
+   * strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    this.eventsListeners_.push(listener);
+  }
+
+  /**
+   * Registers a callback function to be called when extra setup data is
+   * recevied from an AMP Analytics tag.
+   * @param {!Function} listener A function to receive any extra
+   * setup data sent by the creative
+   */
+  registerAmpAnalytics3pExtraDataListener(listener) {
+    this.extraDataListeners_.push(listener);
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @param {!string} message The message to send.
+   */
+  sendMessageToCreative(message) {
+    this.iframeMessagingClient_.sendMessage('ampAnalytics3pResponse',
+      {senderId: this.id_, ampAnalytics3pResponse: message});
+  }
+
+  /**
+   * Gets any optional extra data that should be made available to the
+   * cross-domain frame, in the context of a particular creative.
+   * @param {number} senderId The ID of the creative that sent the extra data
+   * @returns {?string}
+   */
+  getExtraData(senderId) {
+    return this.senderIdToExtraData_[senderId];
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @private
+   */
+  sendReadyMessageToCreative_() {
+    this.iframeMessagingClient_.sendMessage('ampAnalytics3pReady',
+      {senderId: this.id_});
+  }
+
+  /**
+   * Handles messages received from the creative, by enqueueing them to
+   * be sent to the third-party vendor's metrics-collection page
+   * @param {!Object} received The messages that were received from the creative
+   * @private
+   */
+  onReceivedMessagesFromCreative_(received) {
+    if (!received || !received.ampAnalytics3pMessages) {
```

Comment:
I tend to reference object fields that come from outside a binary via string reference to ensure the compiler does not accidentally rename as part of minification, so: received['ampAnalytics3pMessages'].  Or even better would be to create a typedef for the message format and place in an extern to be referenced by sender and receiver binaries to ensure consistent format/typing.